src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Author: Tej Sukhatme
+ */
+
+#include <shogun/features/DotFeatures.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/labels/RegressionLabels.h>
+#include <shogun/lib/observers/ObservedValueTemplated.h>
+#include <shogun/machine/GLM.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/mathematics/NormalDistribution.h>
+#include <shogun/mathematics/RandomNamespace.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+#include <shogun/optimization/SGDMinimizer.h>
+
+#include <cmath>
+
+using namespace shogun;
+
+GLM::GLM()
+{
+	SG_ADD_OPTIONS(
+	    (machine_int_t*)&distribution, "distribution_type",
+	    "variable to store name of distribution type",
+	    ParameterProperties::HYPER, SG_OPTIONS(POISSON));
+	SG_ADD(
+	    &m_eta, "eta",
+	    "threshold parameter that linearizes the exp() function above eta",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_lambda, "lambda", "regularization parameter of penalty term",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_alpha, "alpha",
+	    "weighting between L1 penalty and L2 penalty term of the loss function",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_tolerance, "tolerance", "convergence threshold or stopping criteria",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_learning_rate, "learning_rate", "learning rate for gradient descent",
+	    ParameterProperties::HYPER);
+
+	m_gradient_updater = std::make_shared<GradientDescendUpdater>();
+	m_penalty = std::make_shared<ElasticNetPenalty>();
+	m_cost_function = std::make_shared<GLMCostFunction>();
+}
+
+GLM::GLM(
+    GLM_DISTRIBUTION distr, float64_t alpha, float64_t lambda,
+    float64_t learning_rate, int32_t max_iterations, float64_t tolerance,
+    float64_t eta)
+    : GLM()
+{
+	distribution = distr;
+	m_alpha = alpha;
+	m_lambda = lambda;
+	m_learning_rate = learning_rate;
+	m_max_iterations = max_iterations;
+	m_tolerance = tolerance;
+	m_eta = eta;
+
+	m_penalty->set_l1_ratio(m_alpha);
+}
+
+std::shared_ptr<RegressionLabels>
+GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	if (data)
+	{
+		if (!data->has_property(FP_DOT))
+			error("Specified features are not of type CDotFeatures");
+		set_features(std::static_pointer_cast<DotFeatures>(data));
```

Comment:
is there a reason why the features are added to the state here?

Reply:
No particular reason, I did it this way is it was done like this in LinearMachine:
https://github.com/shogun-toolbox/shogun/blob/ec557d8604b7bf5c2f4ce411aaa2dc0970839d8d/src/shogun/machine/LinearMachine.cpp#L70-L76

Reply:
Should I remove it?

Reply:
I am not sure, you'll have to see if this has side effects that you rely on. @LiuYuHui do you know if this is required? Because you had to refactor this code recently in the feature branch